### ISSUE TITLE
Changing number of AudioDestinationNode outputs from 0 to 1

### ISSUE THREAD

**Issue by olivierthereaux** [WebAudio/web-audio-api#349]
_Reported by ROC in http://lists.w3.org/Archives/Public/public-audio/2014JulSep/0145.html_

This came up in public-media-capture. Suppose you want to capture the output of an AudioContext to a MediaStream or MediaRecorder. The natural thing to do would be something like
```
  var d = context.createMediaStreamDestination();
  context.destination.connect(d);
```
or
```
  var r = new MediaRecorder(context.destination);
```
but that doesn't work because the AudioDestinationNode has no outputs.

So how about we make AudioDestinationNode output the AudioContext output (the mix of its inputs)? Then the above code would work. I'm quite confident it would be a Web-compatible change.

**Comment by hoch** [comment]
Currently nothing prevents the following from happening:

```js
var context = new AudioContext();
var osc = context.createOscillator();
var gain = context.createGain();
osc.connect(gain).connect(context.destination);
context.destination.connect(gain);
```

I am reopening the issue and propose this type of connection is not allowed. The output of destination node can only be connected to MediaRecorder or createMediaStreamDestination.

I must admit this is a short-sighted solution for getting a stream from the graph. We can easily create a gain node just before the destination and intercept the stream from it. A node with a valid output that cannot be connected to nothing but few exceptions is weird.

**Comment by padenot** [comment]
Cycles are already handled fine, I don't think this is an issue. Special casing a particular node would be weird.

Being able to connect the destination to something is nice when you're not in control of the graph, we should not restrict things too much.

**Comment by hoch** [comment]
@padenot Just curious, have you implemented the cycle detection? I've just tested on FF but I don't see any exception being thrown. Have we decided to ignore it silently?

http://webaudio.github.io/web-audio-api/#rendering-loop

**Comment by padenot** [comment]
The `AudioDestinationNode` output is not yet implemented in Firefox.

Cycle detection works my muting cycles. I think it has been like that for a while.

The reason is that we don't want to run the cycle detection algorithm for each `connect()` call from the main thread to be able to synchronously throw an exception.

**Comment by rtoy** [comment]
What does "works my muting cycles" mean?

**Comment by padenot** [comment]
Yeah typo.

I meant to say that, on the rendering thread, if an illegal cycle is detected (i.e. a cycle that does not contain at least a `DelayNode`, the cycle is muted. This can be done as part of pulling the nodes (in chrome), or in Firefox's ordering algorithm. This is therefore folded into other calculations, and happens on the rendering thread.

Throwing an exception would mean doing a topological sort on the main thread for each `connect()` in order to throw the exception, since it has to be synchronous.

iirc, this was changed ages ago when we started the Firefox's implementation.

**Comment by rtoy** [comment]
If this were any other node, I would agree that special casing is to be avoided.  But the `AudioDestinationNode` is already very special. It's the only node that sends audio to the output device and is the only node that captures rendered data for an offline context.

**Comment by rtoy** [comment]
Closing.  Commit c705390bcf6 says it fixed #349, but it didn't.  A little digging says that the number of outputs was actually changed in commit d48c4a54
